ports/stm/common-hal/microcontroller/Pin.c

```diff
@@ -70,6 +70,10 @@ void reset_all_pins(void) {
 
 // Mark pin as free and return it to a quiescent state.
 void reset_pin_number(uint8_t pin_port, uint8_t pin_number) {
+    if ( pin_number == NO_PIN ) {
+        return;
+    }
+
     if (pin_port == 0x0F) {
         return;
     }
```

Comment:
@hierophect can you take a look at the stm parts of this PR please?  (I _suspect_ that the check of pin_port check is the stm equivalent of NO_PIN, except there's an entire 'no port')

Reply:
I do not think that this stm port change is needed.  The current code does not use NO_PIN .  It is coded to use NULL.  NULL is checked before each call to a function that would use it .

Reply:
I'd like to see this in the STM port. It looks like there are places that do not check for a valid pin object: https://github.com/adafruit/circuitpython/blob/main/ports/stm/common-hal/microcontroller/Pin.c#L100

In fact, we should probably make this more resilient by modeling after SAMD which checks validity rather that for the special value.

Reply:
Sorry I'm late to the party here. If I'm understanding this correctly @tannewt you'd rather have functions like common_hal_reset_pin check whether the incoming pin object is NULL? Rather than having the special values?

The main thing the non-object API is used for (port and number) is iterating resets - iterating through the pin objects for a microcontroller is kind of annoying, and doing it with the STM32 port and number values is easier. That said, we have done pin iteration a couple times now, so potentially the entire secondary non-object API could be removed and replaced with an object-only one that checks for pointer validity exclusively.

Reply:
@hierophect Any pin API that takes a pin pointer should check for NULL.

Any API that takes port and pin numbers should validate them. Only checking for a special value risks not catching other invalid values. When we need to set an invalid value a special value is appropriate.